clang/lib/AST/PreorderAST.cpp

```diff
@@ -377,6 +377,127 @@ void PreorderAST::NormalizeExprsWithoutConst(Node *N) {
   AddNode(L, B);
 }
 
+bool PreorderAST::GetDerefOffset(Node *UpperNode, Node *DerefNode,
+				 llvm::APSInt &Offset) {
+  // Extract the offset by which a pointer is dereferenced. For the pointer we
+  // compare the dereference expr with the declared upper bound expr. If the
+  // non-integer parts of the two exprs are not equal we say that a valid
+  // offset does not exist and return false. If the non-integer parts of the
+  // two exprs are equal we require that the integer part of the deref expr be
+  // numerically greater than or equal to the integer part of the upper bound
+  // expr. If this condition is not satisfied we return false. If the condition
+  // is satisfied we return true and the offset in this case is calculated as
+  // (integer part of deref expr - integer part of upper bound expr).
+
+  if (auto *B1 = dyn_cast_or_null<BinaryNode>(UpperNode)) {
+    auto *B2 = dyn_cast_or_null<BinaryNode>(DerefNode);
+
+    // While creating the preorder ASTs we normalize exprs like "*p" to "*(p +
+    // 0)". So we cannot have a case where the UpperNode is a BinaryNode and
+    // the DerefNode is not, or vice versa.
+
+    // Return false if the types of the nodes mismatch.
+    if (!B2)
+      return false;
+
+    // Return false if the opcodes mismatch.
+    // This means the two exprs are like:
+    // (p + i) and (p * i)
+    if (B1->Opc != B2->Opc)
+      return false;
+
+    // Return false if the number of children of the nodes differ.
+    // This means the two exprs are like:
+    // (p + i) and (p + i + j)
+    if (B1->Children.size() != B2->Children.size())
+      return false;
+
+    // Match each child of the two nodes.
+    for (size_t I = 0; I != B1->Children.size(); ++I) {
+      auto *Child1 = B1->Children[I];
+      auto *Child2 = B2->Children[I];
+
+      if (!GetDerefOffset(Child1, Child2, Offset))
```

Comment:
I'm not following why a recursive function call is needed here.   Shouldn't the children of the two nodes be identical, except for the children that are integer offsets?